posthog/hogql_queries/insights/paths_query_runner.py

```diff
@@ -0,0 +1,468 @@
+        timings: Optional[HogQLTimings] = None,
+        modifiers: Optional[HogQLQueryModifiers] = None,
+        limit_context: Optional[LimitContext] = None,
+    ):
+        super().__init__(query, team=team, timings=timings, modifiers=modifiers, limit_context=limit_context)
+
+        if not self.query.pathsFilter:
+            self.query.pathsFilter = PathsFilter()
+
+        self.event_in_session_limit = self.query.pathsFilter.step_limit or EVENT_IN_SESSION_LIMIT_DEFAULT
+
+        self.regex_groupings: list[str] = []
+        if self.query.pathsFilter.path_groupings:
+            self.regex_groupings = [
+                escape(grouping).replace("\\*", ".*") for grouping in self.query.pathsFilter.path_groupings
+            ]
+
+    @property
+    def group_type_index(self) -> int | None:
+        return self.query.aggregation_group_type_index
+
+    def _get_event_query(self) -> list[ast.Expr]:
+        conditions: list[ast.Expr] = []
+        or_conditions: list[ast.Expr] = []
+
+        if not self.query.pathsFilter.include_event_types:
+            return []
+
+        if PathType.field_pageview in self.query.pathsFilter.include_event_types:
+            or_conditions.append(parse_expr(f"event = '{PAGEVIEW_EVENT}'"))
+
+        if PathType.field_screen in self.query.pathsFilter.include_event_types:
+            or_conditions.append(parse_expr(f"event = '{SCREEN_EVENT}'"))
+
+        if PathType.custom_event in self.query.pathsFilter.include_event_types:
+            or_conditions.append(parse_expr("NOT startsWith(events.event, '$')"))
+
+        if or_conditions:
+            conditions.append(ast.Or(exprs=or_conditions))
+
+        if self.query.pathsFilter.exclude_events:
+            conditions.append(
+                ast.CompareOperation(
+                    op=ast.CompareOperationOp.NotIn,
+                    left=ast.Field(chain=["path_item"]),
+                    right=ast.Constant(value=self.query.pathsFilter.exclude_events),
+                )
+            )
+
+        if conditions:
+            return [ast.And(exprs=conditions)]
+
+        return []
+
+    def _should_query_event(self, event: str) -> bool:
+        if not self.query.pathsFilter.include_event_types:  # TODO: include_custom_events ?
+            return event not in (self.query.pathsFilter.exclude_events or [])
+
+        return event in (self.query.pathsFilter.include_event_types or [])
+
+    def construct_event_hogql(self) -> str:
+        event_hogql = "event"
+
+        if self._should_query_event(HOGQL):
+            event_hogql = self.query.pathsFilter.paths_hogql_expression or event_hogql
+
+        if self._should_query_event(PAGEVIEW_EVENT):
+            event_hogql = f"if(event = '{PAGEVIEW_EVENT}', replaceRegexpAll(ifNull(properties.$current_url, ''), '(.)/$', '\\\\1'), {event_hogql})"
+
+        if self._should_query_event(SCREEN_EVENT):
+            event_hogql = f"if(event = '{SCREEN_EVENT}', properties.$screen_name, {event_hogql})"
```

Comment:
Same comment about f-strings here.

Also, quick shortcut if you want it: `ifNull(properties.$current_url, '')` can also be written with the JS style nullish coalescing operator: `properties.$current_url ?? ''`